File: frontend/admin/js/production.js
/* ═══════════════ PRODUCTION ═══════════════
   The order list answers "what did this customer buy". This answers the
   opposite question: what to pull off the shelf and put on the press,
   batched across every open order. */

async function renderProduction(el, sub){
  const tabs = [['print-jobs','Print jobs'],['artwork-files','Artwork files'],
                ['proof-log','Proof log']];
  el.innerHTML = `<div class="row" style="margin-bottom:18px">
    ${tabs.map(([k,label])=>`<button class="btn btn-sm ${(sub||'print-jobs')===k?'btn-primary':'btn-quiet'}"
      onclick="goTo('production','${k}')">${esc(label)}</button>`).join('')}
  </div><div id="prodBody"></div>`;
  const body = document.getElementById('prodBody');
  if(sub === 'artwork-files') return renderArtwork(body);
  if(sub === 'proof-log')     return renderProofs(body);
  return renderJobs(body);
}

/* ── Print jobs ──────────────────────────────────────────────── */
async function renderJobs(el){
  el.innerHTML = '<div class="empty">Loading…</div>';
  const d = await api('/api/admin/production/jobs');
  if(!d.ok){ el.innerHTML = `<div class="empty">${esc(d.error)}</div>`; return; }
  if(!d.jobs.length){
    el.innerHTML = '<div class="empty"><span class="material-symbols-outlined">print</span>' +
      '<br>Nothing on the floor. Orders appear here once their proof is approved.</div>';
    return;
  }
  el.innerHTML = `
    <div class="grid g3" style="margin-bottom:18px">
      <div class="stat"><span>Blanks to pull</span><b>${d.total_pieces}</b></div>
      <div class="stat"><span>Groups</span><b>${d.jobs.length}</b></div>
      <div class="stat"><span>Orders involved</span><b>${
        new Set(d.jobs.flatMap(j=>j.lines.map(l=>l.order))).size}</b></div>
    </div>
    ${d.jobs.map(j=>`
      <div class="card">
        <div class="spread" style="margin-bottom:12px">
          <div>
            <h2><span class="swatch-dot" style="background:${esc(j.color_hex)}"></span>
              ${esc(j.product)} — ${esc(j.color_name)}</h2>
            <p class="tiny muted">${j.lines.length} order line(s) on the same blank</p>
          </div>
          <div class="stat" style="padding:8px 14px"><span>Pull</span><b>${j.qty}</b></div>
        </div>
        <table>
          <thead><tr><th>Order</th><th>Sizes</th><th>Stage</th><th>Artwork</th></tr></thead>
          <tbody>${j.lines.map(l=>`<tr>
            <td data-label="Order"><b>${esc(l.order)}</b><br>
              <span class="tiny dim">${esc(l.customer)} · ${esc(fmtWhen(l.created))}</span></td>
            <td data-label="Sizes">${Object.entries(l.sizes||{}).filter(([,n])=>+n>0)
              .map(([k,n])=>`<span class="size-chip"><b>${n}</b>×${esc(k)}</span>`).join('')
              || `<span class="tiny dim">${l.qty} pcs</span>`}</td>
            <td data-label="Stage">${stageBadge(l.status)}</td>
            <td data-label="Artwork">${['front','back'].map(side=>{
              const url = l.art && l.art[side];
              const spec = l.spec && l.spec[side];
              if(!url && !spec) return '';
              return `<div style="margin-bottom:6px">
                ${url?`<a class="btn btn-quiet btn-sm" href="${esc(url)}" download>
                  <span class="material-symbols-outlined" style="font-size:15px">download</span>
                  ${side}</a>`:`<span class="tiny dim">${side}: no file</span>`}
                ${spec?`<div class="tiny dim">${(spec.layers||[]).map(x=>
                  `${x.w}×${x.h} cm, ${x.fromTop} from top`).join('; ')}</div>`:''}
              </div>`;
            }).join('') || '<span class="tiny dim">—</span>'}</td>
          </tr>`).join('')}</tbody>
        </table>
      </div>`).join('')}`;
}

/* ── Artwork files ───────────────────────────────────────────── */
async function renderArtwork(el){
  el.innerHTML = '<div class="empty">Loading…</div>';
  const d = await api('/api/admin/production/artwork');
  if(!d.ok){ el.innerHTML = `<div class="empty">${esc(d.error)}</div>`; return; }
  const mb = (d.total_bytes/1048576).toFixed(1);
  el.innerHTML = `
    <div class="grid g3" style="margin-bottom:18px">
      <div class="stat"><span>Print files</span><b>${d.files.length}</b></div>
      <div class="stat"><span>On the volume</span><b>${mb} MB</b></div>
      <div class="stat"><span>Orphaned</span><b>${d.orphans}</b></div>
    </div>
    ${d.orphans ? `<div class="alert warn" style="margin-bottom:16px">
      <span class="material-symbols-outlined">warning</span>
      <span><b>${d.orphans} file(s) belong to no order.</b> Those come from a
        checkout that uploaded artwork and then failed. Harmless, but they're
        the only thing on the volume that grows without being cleaned up.</span>
    </div>` : ''}
    <div class="card" style="padding:0;overflow:hidden">
      ${d.files.length ? `<table>
        <thead><tr><th>File</th><th>Order</th><th>Side</th><th>Format</th><th>Size</th><th></th></tr></thead>
        <tbody>${d.files.map(f=>`<tr class="${f.orphan||f.cancelled?'row-off':''}">
          <td data-label="File"><span class="tiny" style="font-family:monospace">${esc(f.name)}</span></td>
          <td data-label="Order">${f.order
            ? esc(f.order) + (f.cancelled?' <span class="badge badge-quiet">cancelled</span>':'')
            : '<span class="badge badge-warn">orphan</span>'}
            ${f.product?`<br><span class="tiny dim">${esc(f.product)}</span>`:''}</td>
          <td data-label="Side">${esc(f.side||'—')}</td>
          <td data-label="Format"><span class="badge badge-quiet">${esc(f.format)}</span></td>
          <td data-label="Size">${(f.bytes/1024).toFixed(0)} KB</td>
          <td data-label=""><a class="btn btn-quiet btn-sm" href="${esc(f.url)}" download>
            <span class="material-symbols-outlined" style="font-size:15px">download</span></a></td>
        </tr>`).join('')}</tbody></table>`
        : '<div class="empty">No print files yet.</div>'}
    </div>
    <p class="tiny dim" style="margin-top:12px">
      These are the only copies production can print from — the order itself
      stores a preview thumbnail, not the artwork.
    </p>`;
}

/* ── Proof log ───────────────────────────────────────────────── */
async function renderProofs(el){
  el.innerHTML = '<div class="empty">Loading…</div>';
  const d = await api('/api/admin/production/proofs');
  if(!d.ok){ el.innerHTML = `<div class="empty">${esc(d.error)}</div>`; return; }
  el.innerHTML = `
    ${d.overdue ? `<div class="alert warn" style="margin-bottom:16px">
      <span class="material-symbols-outlined">schedule</span>
      <span><b>${d.overdue} order(s) past the ${d.sla_hours}-hour proof promise.</b>
        The storefront tells customers they'll get a digital proof within
        ${d.sla_hours} hours — these are already late.</span>
    </div>` : ''}
    <div class="card">
      <h2>Waiting on approval</h2>
      <p class="tiny muted">Orders at the Proof sent stage, oldest first.</p>
      ${d.waiting.length ? `<table>
        <thead><tr><th>Order</th><th>Customer</th><th>Waiting</th><th></th></tr></thead>
        <tbody>${d.waiting.map(w=>`<tr>
          <td data-label="Order"><b>${esc(w.id)}</b></td>
          <td data-label="Customer">${esc(w.customer)}<br>
            <span class="tiny dim">${esc(w.email)}</span></td>
          <td data-label="Waiting"><b style="${w.overdue?'color:var(--pink-ink)':''}">
            ${w.hours} h</b>${w.overdue?' <span class="badge badge-warn">overdue</span>':''}</td>
          <td data-label=""><button class="btn btn-quiet btn-sm"
            onclick="goTo('orders','awaiting-proof')">Open in Orders</button></td>
        </tr>`).join('')}</tbody></table>`
        : '<div class="empty">Nothing waiting on a proof.</div>'}
    </div>

    <div class="card">
      <h2>Recently approved</h2>
      <p class="tiny muted">Taken from the audit log — a Proof sent → Approved
        move is exactly "the proof was accepted".</p>
      ${d.approved.length ? `<table><tbody>${d.approved.map(a=>`<tr>
        <td data-label="Order"><b>${esc(a.id)}</b></td>
        <td data-label="When" class="tiny dim">${esc(fmtWhen(a.created))}
          ${a.actor?' · '+esc(a.actor):''}</td>
      </tr>`).join('')}</tbody></table>`
        : '<div class="empty">No approvals recorded yet.</div>'}
    </div>

    <p class="tiny dim">
      Proofs are marked approved by staff here — there's no email transport
      yet, so nothing is sent to the customer automatically.
    </p>`;
}

File: backend/production.py
# ── Production: print queue, artwork files, proof tracking ──────
# The order list answers "what did this customer buy". The print floor needs
# the opposite view: what to pull off the shelf and what to put on the press,
# batched across every open order.
import json
import os

from flask import Blueprint, jsonify, request

from db import get_db
from permissions import require
from artwork import ART_DIR

production_bp = Blueprint("production", __name__, url_prefix="/api/admin/production")

# Stages a garment is actually being made in. 0 is waiting on the customer's
# proof approval and 4/5 have left the building, so neither belongs in a
# queue for the press.
IN_PRODUCTION = (1, 2, 3)

# The storefront promises a digital proof within 2 hours. Anything sitting
# unapproved past that is a broken promise, so the panel says so rather than
# leaving it to be noticed.
PROOF_SLA_HOURS = 2


def _open_orders(db, stages=IN_PRODUCTION):
    marks = ",".join("?" * len(stages))
    return db.execute(
        """SELECT o.*, u.name AS customer FROM orders o
           JOIN users u ON u.id=o.user_id
           WHERE o.cancelled=0 AND o.payment_status!='pending' AND o.status IN (%s)
           ORDER BY o.created""" % marks, stages).fetchall()


@production_bp.route("/jobs")
@require("production")
def print_jobs():
    """Everything on the floor, grouped by blank: product + garment colour.

    Grouping by artwork would be pointless — every design is bespoke, so
    each would be its own group. Grouping by the blank is what lets someone
    pull twelve green polos once instead of walking to the shelf per order.
    """
    db = get_db()
    colors = {r["color_hex"].upper(): r["color_name"] for r in db.execute(
        "SELECT DISTINCT color_hex,color_name FROM variants")}

    groups = {}
    for o in _open_orders(db):
        try:
            items = json.loads(o["items_json"])
        except (ValueError, TypeError):
            continue
        for it in items if isinstance(items, list) else []:
            if not isinstance(it, dict):
                continue
            hexv = (it.get("shirt") or "").upper()
            key = (it.get("pid") or "?", hexv)
            g = groups.setdefault(key, {
                "pid": key[0], "product": it.get("product") or key[0],
                "color_hex": hexv, "color_name": colors.get(hexv, hexv),
                "qty": 0, "lines": [],
            })
            sizes = it.get("sizes") if isinstance(it.get("sizes"), dict) else {}
            qty = sum(n for n in sizes.values() if isinstance(n, int) and n > 0) \
                or (it.get("qty") if isinstance(it.get("qty"), int) else 0)
            g["qty"] += qty
            g["lines"].append({
                "order": "PL-" + str(1000 + o["id"]),
                "customer": o["customer"],
                "status": o["status"],
                "created": o["created"],
                "sizes": sizes,
                "qty": qty,
                "art": it.get("art") or {},
                "spec": it.get("spec") or {},
                "thumb": it.get("thumb"),
            })
    out = sorted(groups.values(), key=lambda g: (-g["qty"], g["product"]))
    return jsonify(ok=True, jobs=out,
                   total_pieces=sum(g["qty"] for g in out))


@production_bp.route("/artwork")
@require("production")
def artwork_files():
    """Every print file, matched back to the order that produced it.

    Also reports orphans — files on the volume that no order references.
    Those come from a checkout that uploaded artwork and then failed, and
    they're the only thing on the volume that grows without bound.
    """
    db = get_db()
    used = {}
    for o in db.execute("SELECT id,items_json,cancelled,status,created FROM orders"):
        try:
            items = json.loads(o["items_json"])
        except (ValueError, TypeError):
            continue
        for it in items if isinstance(items, list) else []:
            if not isinstance(it, dict):
                continue
            for side, url in (it.get("art") or {}).items():
                if not isinstance(url, str):
                    continue
                used[os.path.basename(url)] = {
                    "url": url, "side": side,
                    "order": "PL-" + str(1000 + o["id"]),
                    "product": it.get("product"),
                    "cancelled": bool(o["cancelled"]),
                    "created": o["created"],
                }

    files, orphans, total = [], 0, 0
    try:
        names = sorted(os.listdir(ART_DIR))
    except OSError:
        names = []
    for name in names:
        path = os.path.join(ART_DIR, name)
        if not os.path.isfile(path):
            continue
        size = os.path.getsize(path)
        total += size
        meta = used.get(name)
        if not meta:
            orphans += 1
        files.append({
            "name": name, "bytes": size,
            "format": name.rsplit(".", 1)[-1].lower(),
            "url": "/api/artwork/" + name,
            "order": meta["order"] if meta else None,
            "side": meta["side"] if meta else None,
            "product": meta["product"] if meta else None,
            "cancelled": meta["cancelled"] if meta else False,
            "orphan": meta is None,
        })
    files.sort(key=lambda f: (f["order"] or "", f["name"]), reverse=True)
    return jsonify(ok=True, files=files, orphans=orphans, total_bytes=total)


@production_bp.route("/proofs")
@require("production")
def proofs():
    """Orders waiting on proof approval, oldest first, with the ones past
    the 2-hour promise flagged. Approval history comes from the audit log —
    a 0 -> 1 stage change is exactly "customer approved the proof"."""
    db = get_db()
    waiting = []
    for o in db.execute("""SELECT o.*, u.name AS customer, u.email FROM orders o
                           JOIN users u ON u.id=o.user_id
                           WHERE o.cancelled=0 AND o.payment_status!='pending' AND o.status=0
                           ORDER BY o.created"""):
        age = db.execute(
            "SELECT (julianday('now') - julianday(?)) * 24 AS h", (o["created"],)
        ).fetchone()["h"]
        waiting.append({
            "id": "PL-" + str(1000 + o["id"]), "customer": o["customer"],
            "email": o["email"], "created": o["created"],
            "hours": round(age, 1), "overdue": age > PROOF_SLA_HOURS,
        })
    approved = [{
        "id": "PL-" + str(1000 + r["entity_id"]), "created": r["created"],
        "actor": r["actor_name"],
    } for r in db.execute("""
        SELECT a.*, u.name AS actor_name FROM audit_log a
        LEFT JOIN users u ON u.id=a.actor_id
        WHERE a.entity_type='order' AND a.action='stage'
          AND a.from_val='0' AND a.to_val='1'
        ORDER BY a.id DESC LIMIT 30""")]
    return jsonify(ok=True, waiting=waiting, approved=approved,
                   sla_hours=PROOF_SLA_HOURS,
                   overdue=sum(1 for w in waiting if w["overdue"]))
